### PR TITLE
Introduce public setter for Activity 

### PR DESCRIPTION
**Pull request by lmolkova**
See #29207

### CODE REVIEW

**Comment by vancem** [comment]
I will take a look today.

File: src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs [hunk 1]
```diff
@@ -446,6 +448,17 @@ private static unsafe long GetRandomNumber()
             return *((long*)&g);
         }
 
+        private static bool ValidateSetCurrent(Activity activity)
+        {
+            bool canSet = activity == null || (activity.Id != null && !activity.isFinished);
```
Comment (vancem) [inline src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs]: There are potential races here, but I believe the right answer is to simply tolerate them.

**Comment by lmolkova** [comment]
@dotnet-bot help

File: src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs (hunk 1 above)
Comment (lmolkova) [reply src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs]: Agree, concurrent usage and modification of `Activity.Current` should be extremely rare and does not look like a valid scenario

**Comment by lmolkova** [comment]
@dotnet-bot test OSX x64 Debug Build

**Comment by lmolkova** [comment]
@dotnet-bot test Packaging All Configurations x64 Debug Build

**Comment by lmolkova** [comment]
thanks for the review, @vancem!

**Pull request merged**